### ISSUE TITLE
Multiple traefik containers targeting same external database causing HasServerCrashed

### ISSUE THREAD

**Issue by sirhc101** [microsoft/navcontainerhelper#1867]
**Describe the issue**
We have setup two containers targeting the same external database. Both containers using traefik. Both containers are working pretty well if the other one is not running/started. As soon as we start the other container we encounter the following behavior:

- Session are timed out very fast. If you do not work for a few seconds you're kicked out 
- In event log we see this warning almost every 30 seconds:
![image](https://user-images.githubusercontent.com/39621236/113696320-213ee280-96d2-11eb-8bf6-a1c4c757f170.png)

I searched a lot and cannot find any reason. As I said when running just one container everything works well. Unfortunately we need both containers due to the fact AL language extension in VSCode is not supporting AAD auth. for other the `businesscentral.dynamics.com` sandboxes. So the second container is using `UserPassword` auth.

I've just one idea: In SQL Server Management Studio I've seen that both containers are using the same hostname. They both using the same server instance (BC). So I assume that the problem is caused by the "clean-up" or "health check" of server instances. In `[Server Instance]` table in SQL Server I see both server instances, but one does have the status "2"
![image](https://user-images.githubusercontent.com/39621236/113697188-28b2bb80-96d3-11eb-8785-55d9b0a29c27.png)

I saw in NavContainerHelper code that, when using traefik, the hostname is always the first part from the public dns name (subdomain), but is there are reason why the hostname of the computer needs to be this name? I tried `Rename-Computer` but of course it doesn't work because everytime the container restarts the hostname is set again.

Is there a way to change this or do you think I'm completely wrong and the problem is somewhere else?

Thank you in advance.

**Comment by freddydk** [comment]
This could very well be due to the hostname.
You could try to modify the line setting the hostname to traefikhostname - to use the containername instead.
I am not entirely sure what impact this has and why the traefik hostname is set as it is.
Seems strange that all containers in a traefik environment will have the same hostname
Maybe @tfenster recalls the reason why?

**Comment by sirhc101** [comment]
Update: I've added a new subdomain, so the containers do now get different hostnames and the problem seems to be gone. I'll try to modify the `New-NavContainer` code a bit to figure out how we can handle this a little better.

**Comment by tfenster** [comment]
> This could very well be due to the hostname.
> You could try to modify the line setting the hostname to traefikhostname - to use the containername instead.
> I am not entirely sure what impact this has and why the traefik hostname is set as it is.
> Seems strange that all containers in a traefik environment will have the same hostname
> Maybe @tfenster recalls the reason why?

Yes, the WebClient and/or NST didn't work correctly if the hostname wasn't set to the same value as came in from the external request

**Comment by sirhc101** [comment]
With commit [96f8503](https://github.com/sirhc101/navcontainerhelper/commit/96f85035c26456fd0504fb8bc4441caf8610932a) I simply removed the `--hostname $traefikHostname` from `$additionalParameters`. Now the hostname of the container is equal to the container name (like how it is w/o usage of traefik). For now I cannot find any weird behavior or similar. I will test a bit in my environments but looks good.

@tfenster Can you remember what exactly didn't worked?

**Comment by freddydk** [comment]
I am open for a PR on this.
I would like to have a config setting called TraefikUseDnsNameAsHostName, which people can set to true to get the old behavior.
Config settings are defined in BcContainerHelper.psm1 - and should be straightforward to just add a new.
Default should be false - meaning that we get the new behavior.

**Comment by tfenster** [comment]
@freddydk that would be great. Btw, the PR that we needed to make traefik 2 work completely for BC is merged and should appear with the next feature release (https://github.com/traefik/traefik/pull/7645), so that I can then also do a PR for bcch

**Comment by tfenster** [comment]
I'll try :)

**Comment by sirhc101** [comment]
@freddydk Regarding the upgrade from traefic v1 to v2. I've added a command [`Add-DomainToTraefikConfig`](https://github.com/sirhc101/navcontainerhelper/commit/8e1d9aa0c3109765b1d6a6787e3c68321c0ba85d) to support multiple domains (or subdomains) for acme challenge. Initially this was a workaround to be able to have two containers with different names but I think it can be useful. It extends the traefik configuration so get a valid certificate.
When upgrading to traefik v2 this must be updated as well, because v2 uses yaml as configuration.

Regarding the configuration. I will add the `TraefikUseDnsNameAsHostName` setting and create the PR after finishing my tests

**Comment by sirhc101** [comment]
@tfenster I looked at the [traefik documentation for HTTP challenge](https://doc.traefik.io/traefik/https/acme/#httpchallenge) but to be honest... i've no clue how to configure similar to traefik v1.
In traefik v1 it's pretty simple:
```toml
[..]
[acme.httpChallenge]
entryPoint = "http"
[[acme.domains]]
   main = "domain.com"
   sans = ["sub1.domain.com","sub2.domain.com"]
[..]
```

Could you describe how to do the same in traefik v2 yaml configuration? Then I would to extend `Add-DomainToTraefikConfig` so it's ready for v1 and v2.

**Comment by tfenster** [comment]
@sirhc101 v2 automatically picks the domain name up from the labels. So if you have a rule like e.g.
```
Host(`sub1.domain.com`) && PathPrefix(`/489b16f041cd`)
```
it would automatically request the cert for that sub

**Comment by sirhc101** [comment]
> @sirhc101 v2 automatically picks the domain name up from the labels. So if you have a rule like e.g.
> 
> ```
> Host(`sub1.domain.com`) && PathPrefix(`/489b16f041cd`)
> ```
> 
> it would automatically request the cert for that sub

@tfenster Thank you for your feedback.

@freddydk In this case you just have to disable `Add-DomaintToTraefikConfig` when using traefik v2. It's only relevant for v1

**Comment by sirhc101** [comment]
Closed with #1872

**Comment by KristofKlein** [comment]
Sorry for a comment on a closed issue that is really old - but as I was playing around with this things I wanted to share my findings as well:
I was doing some lab work around running containers against an external DB as well. And at one point I was playing around with traefik as well ( working config for v3!) but ended up with also an unstable solution. So I found this post and I came to the same result, while utilizing one more component here: gMSA. So I was running with a docker compose file to get me all sorts of BC container started. It ended up to be a super unstable solution the moment I scaled things. (in fact the very first second container already made things bad). So I used individual hostnames, containers are all up and running. But me also getting this Healthcheck issue and kicked sessions. in the end it turned out, that, no matter what you do to the hostname, gMSA will set them all - externally - to the same servername. So I also ended up with what you have in picture two ! The "same" Server, running the "same" Service. 
![image](https://github.com/user-attachments/assets/b86b66d1-3f7c-4bfd-9afd-c188808030d0)

So if you try this "at home" gMSA is a "full stop" for this right now.
